onedal/decomposition/pca.py

```diff
@@ -15,30 +15,77 @@
 # ==============================================================================
 
 import numpy as np
+from sklearn.decomposition._pca import _infer_dimension
+from sklearn.utils.extmath import stable_cumsum
 
 from daal4py.sklearn._utils import sklearn_check_version
 
 from ..common._base import BaseEstimator
-from ..common.hyperparameters import get_hyperparameters
 from ..datatypes import _convert_to_supported, from_table, to_table
 
 
 class PCA(BaseEstimator):
     def __init__(
-        self, n_components=None, is_deterministic=True, method="precomputed", copy=True
+        self,
+        n_components=None,
+        is_deterministic=True,
+        method="cov",
+        whiten=False,
     ):
         self.n_components = n_components
         self.method = method
         self.is_deterministic = is_deterministic
+        self.whiten = whiten
 
-    def get_onedal_params(self, data):
+    def get_onedal_params(self, data, stage=None):
+        if stage is None:
+            n_components = self._resolve_n_components_for_training(data.shape)
+        elif stage == "predict":
+            n_components = self.n_components_
         return {
             "fptype": "float" if data.dtype == np.float32 else "double",
             "method": self.method,
-            "n_components": self.n_components,
+            "n_components": n_components,
             "is_deterministic": self.is_deterministic,
+            "whiten": self.whiten,
         }
 
+    def _resolve_n_components_for_training(self, shape_tuple):
+        if self.n_components is None or self.n_components == "mle":
+            return min(shape_tuple)
+        elif (
+            isinstance(self.n_components, float)
+            and self.n_components > 0.0
+            and self.n_components <= 1.0
+        ):
+            return min(shape_tuple)
+        else:
+            return self.n_components
+
+    def _resolve_n_components_for_result(self, shape_tuple):
+        if self.n_components is None:
+            return min(shape_tuple)
+        elif self.n_components == "mle":
+            return _infer_dimension(self.explained_variance_, shape_tuple[0])
+        elif 0.0 < self.n_components < 1.0:
+            ratio_cumsum = stable_cumsum(self.explained_variance_ratio_)
+            return np.searchsorted(ratio_cumsum, self.n_components, side="right") + 1
+        elif isinstance(self.n_components, float) and self.n_components == 1.0:
+            return min(shape_tuple)
+        else:
+            return self.n_components
```

Comment:
Some handling of invalid n_components here would be nice - currently the error message is ugly (`RuntimeError: Unable to cast Python instance of type <class 'str'> to C++ type '?' (#define PYBIND11_DETAILED_ERROR_MESSAGES or compile in debug mode for details)`) - sklearn and our sklearnex interfaces use `InvalidParameterError: The 'n_components' parameter of PCA must be an int in the range [0, inf), a float in the range (0.0, 1.0), a str among {'mle'} or None. Got 1.5 instead.`